dask/dataframe/io/parquet/arrow.py

```diff
@@ -698,10 +422,571 @@ def read_metadata(
+                        continue  # This partition was filtered
+                    part = {
+                        "piece": (full_path, rg_list, pkeys),
+                        "kwargs": {
+                            "partitions": partition_obj,
+                            "categories": categories,
+                            "filters": filters,
+                            "schema": schema,
+                        },
+                    }
+                    parts.append(part)
+                    if gather_statistics:
+                        stat = cls._aggregate_stats(
+                            filename,
+                            file_row_group_stats[filename][i:i_end],
+                            file_row_group_column_stats[filename][i:i_end],
+                            stat_col_indices,
+                        )
+                        stats.append(stat)
+        else:
+            for filename, row_groups in file_row_groups.items():
+                # Get full path (empty strings should be ignored)
+                full_path = fs.sep.join([p for p in [data_path, filename] if p != ""])
+                pkeys = partition_keys.get(full_path, None)
+                if partition_obj and pkeys is None:
+                    continue  # This partition was filtered
+                rgs = None if use_legacy_dataset else row_groups
+                part = {
+                    "piece": (full_path, rgs, pkeys),
+                    "kwargs": {
+                        "partitions": partition_obj,
+                        "categories": categories,
+                        "filters": filters,
+                        "schema": schema,
+                    },
+                }
+                parts.append(part)
+                if gather_statistics:
+                    stat = cls._aggregate_stats(
+                        filename,
+                        file_row_group_stats[filename],
+                        file_row_group_column_stats[filename],
+                        stat_col_indices,
+                    )
+                    stats.append(stat)
+
+        return parts, stats
+
+    @classmethod
+    def _process_metadata(
+        cls,
+        metadata,
+        single_rg_parts,
+        gather_statistics,
+        stat_col_indices,
+        no_filters,
+        fragment_row_groups,
+    ):
+        """Process row-groups and statistics.
+
+        This method is overridden in `ArrowLegacyEngine`.
+        """
+
+        # Get the number of row groups per file
+        file_row_groups = defaultdict(list)
+        file_row_group_stats = defaultdict(list)
+        file_row_group_column_stats = defaultdict(list)
+        cmax_last = {}
+        for frag in metadata:
+            for row_group in frag.row_groups:
+                statistics = row_group.statistics
+                fpath = frag.path
+                if fragment_row_groups:
+                    file_row_groups[fpath].append(frag)
```

Comment:
Do we know at this point that the fragment consists of only a single row group (so already splitted by row groups)? Because you are looping through the row groups of the fragment, but always assigning the same fragment

Reply:
Thanks for pointing this out.  This actually motivated me to head down yet another rabbit hole with this PR...  So far, I have been using `split_by_row_group` to construnct a list of row-group fragments.  I now realize it makes more sense to keep a list of file fragments, and simply recreate the fragments when (1) filters are applied and (2) at read time.  By "re-create", I am refering to the `make_fragment` tip you shared [below](https://github.com/dask/dask/pull/6534#issuecomment-699113894).